airbyte-integrations/bases/base-typing-deduping-test/src/main/java/io/airbyte/integrations/base/destination/typing_deduping/RecordDiffer.java

```diff
@@ -0,0 +1,386 @@
+   * JSON null is represented as a NullNode. For example, in the JSON blob {"name": null}, the `name`
+   * field is a JSON null, and the `address` field is a SQL null.
+   */
+  public void verifySyncResult(List<JsonNode> expectedRawRecords,
+                               List<JsonNode> actualRawRecords,
+                               List<JsonNode> expectedFinalRecords,
+                               List<JsonNode> actualFinalRecords) {
+    assertAll(
+        () -> diffRawTableRecords(expectedRawRecords, actualRawRecords),
+        () -> diffFinalTableRecords(expectedFinalRecords, actualFinalRecords));
+  }
+
+  public void diffRawTableRecords(List<JsonNode> expectedRecords, List<JsonNode> actualRecords) {
+    String diff = diffRecords(
+        expectedRecords,
+        actualRecords,
+        rawRecordIdentityComparator,
+        rawRecordSortComparator,
+        rawRecordIdentityExtractor,
+        true);
+
+    if (!diff.isEmpty()) {
+      fail("Raw table was incorrect.\n" + diff);
+    }
+  }
+
+  public void diffFinalTableRecords(List<JsonNode> expectedRecords, List<JsonNode> actualRecords) {
+    String diff = diffRecords(
+        expectedRecords,
+        actualRecords,
+        finalRecordIdentityComparator,
+        finalRecordSortComparator,
+        finalRecordIdentityExtractor,
+        false);
+
+    if (!diff.isEmpty()) {
+      fail("Final table was incorrect.\n" + diff);
+    }
+  }
+
+  private static String getPrintableFieldIfPresent(JsonNode record, String field) {
+    if (record.has(field)) {
+      return field + "=" + record.get(field);
+    } else {
+      return "";
+    }
+  }
+
+  /**
+   * Generate a human-readable diff between the two lists. Assumes (in general) that two records with
+   * the same PK, cursor, and extracted_at are the same record.
+   * <p>
+   * Verifies that all values specified in the expected records are correct (_including_ raw_id), and
+   * that no other fields are present (except for loaded_at and raw_id). We assume that it's
+   * impossible to verify loaded_at, since it's generated dynamically; however, we do provide the
+   * ability to assert on the exact raw_id if desired; we simply assume that raw_id is always expected
+   * to be present.
+   *
+   * @param identityComparator Returns 0 iff two records are the "same" record (i.e. have the same
+   *        PK+cursor+extracted_at)
+   * @param sortComparator Behaves identically to identityComparator, but if two records are the same,
+   *        breaks that tie using _airbyte_raw_id
+   * @param recordIdExtractor Dump the record's PK+cursor+extracted_at into a human-readable string
+   * @param extractRawData Whether to look inside the _airbyte_data column and diff its subfields
+   * @return The diff, or empty string if there were no differences
+   */
+  private static String diffRecords(List<JsonNode> originalExpectedRecords,
```

Comment:
this method is doing too much, I think it needs some helper methods. General guidance would be to start with conditionals and extract there since there is readability value alone in having a "operator method" which has nothing but the conditional logic calling out to other methods.

Reply:
fair, will work on this later today/tomorrow.

Reply:
https://github.com/airbytehq/airbyte/pull/27852/files/58e5d10338667f2625879380d9da9b07bcc3b7d8..7539094d533a0432bcc82ed5030bd1d27abacd7c

* unified raw+final record paths. Raw records now just lift the _airbyte_data fields to the root level before comparison
* diffRecords is just iterating through the expected/actual records
* diffSingleRecord does the actual field-level diff
* areJsonNodesEquivalent - I think the boolean flipping broke null-handling; fix it here.